src/test/java/unit/GcsBucketLifecycle.java

```diff
@@ -0,0 +1,390 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import bio.terra.cli.serialization.userfacing.UFAuthStatus;
+import com.google.api.client.util.DateTime;
+import com.google.cloud.Identity;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.StorageClass;
+import harness.TestCommand;
+import harness.baseclasses.SingleWorkspaceUnit;
+import harness.utils.ExternalGCSBuckets;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+/** Tests for specifying lifecycle rules for controlled GCS buckets. */
+@Tag("unit")
+public class GcsBucketLifecycle extends SingleWorkspaceUnit {
+  // external bucket to use for testing the JSON format against GCS directly
+  private Bucket externalBucket;
+
+  @Override
+  @BeforeAll
+  protected void setupOnce() throws IOException {
+    super.setupOnce();
+    externalBucket = ExternalGCSBuckets.createBucket();
+
+    // `terra auth status --format=json`
+    UFAuthStatus authStatus =
+        TestCommand.runAndParseCommandExpectSuccess(UFAuthStatus.class, "auth", "status");
+
+    // grant the user's proxy group write access to the bucket, so we can test calling `terra gsutil
+    // lifecycle` with the same JSON format used for creating controlled bucket resources with
+    // lifecycle rules
+    ExternalGCSBuckets.grantWriteAccess(externalBucket, Identity.group(authStatus.proxyGroupEmail));
+  }
+
+  @Override
+  @AfterAll
+  protected void cleanupOnce() throws IOException {
+    super.cleanupOnce();
+    ExternalGCSBuckets.getStorageClient().delete(externalBucket.getName());
+    externalBucket = null;
+  }
+
+  @Override
+  @BeforeEach
+  protected void setupEachTime() throws IOException {
+    super.setupEachTime();
+
+    workspaceCreator.login();
+
+    // `terra workspace set --id=$id`
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + getWorkspaceId());
+  }
+
+  @Test
+  @DisplayName("lifecycle action delete (condition age)")
+  void deleteAction() throws IOException {
+    String name = "delete_age";
+    BucketInfo.LifecycleRule lifecycleRuleFromGCS =
+        createBucketWithOneLifecycleRule(name, name + ".json");
+
+    expectActionDelete(lifecycleRuleFromGCS);
+    assertEquals(365, lifecycleRuleFromGCS.getCondition().getAge(), "condition age matches");
+  }
+
+  @Test
+  @DisplayName("lifecycle action set storage class (condition age)")
+  void setStorageClassAction() throws IOException {
```

Comment:
nit: It feels like overkill to create a new bucket for every different condition/action. I think this would be fine to have as a unit test on `WorkspaceManagerService#fromCLIObject` that we can translate a CLI input format -> WM API format, but going end-to-end for every variation seems like overkill where we're testing WM functionality more than the CLIs. If we're using the WM API, we don't need to verify that every one of its options does what's promised, one end2end on the lifecycle seems sufficient. Or, why don't we need to verify every condition structure also works as a gsutil format?

Reply:
I think the CLI test should cover deserializing the lifecycle JSON file into a CLI object and converting that into a WSM object, since those are the parts of the e2e that the CLI owns. The deserializing happens in the command class, so these tests should go through there. From the perspective of testing the CLI, I would be happy with a test that stopped just before making the WSM call, but don't see how to do that without modifying the codebase -- it's simpler to go e2e.

The goal of testing each option wasn't to test WSM (though that did end up being useful -- this test showed that 4 options in WSM's public API are actually not supported yet). It was to test the CLI deserialization and conversion into a WSM object.

I included a single test for `gsutil` just to check that the CLI JSON format was readable by GCS directly. The other tests already covered that the CLI deserialization worked.